### PR TITLE
[MMSYS] Improve sound test button

### PR DESCRIPTION
**Pull request by TAN-Gaming**
## Proposed changes
- Make sound test button able to stop the sound (like WinXP's mmsys.cpl)

https://user-images.githubusercontent.com/86486978/201700271-4ecb66b0-d115-464a-8d2b-945394ca2f78.mp4

### CODE REVIEW

**Comment by JoachimHenze** [comment]
Iirc Windows doesn't use real icons for play and stop, but uses some Marlett characters on that button.

**Comment by TAN-Gaming** [comment]
> Iirc Windows doesn't use real icons for play and stop, but uses some Marlett characters on that button. We should do that too!

OK, I'll do that in the later PR.

**Comment by HBelusca** [comment]
> > Iirc Windows doesn't use real icons for play and stop, but uses some Marlett characters on that button. We should do that too!
> 
> OK, I'll do that in the later PR.

This has to be confirmed first, though...

File: dll/cpl/mmsys/sounds.c [hunk 1]
```diff
@@ -1140,9 +1228,19 @@ SoundsDlgProc(HWND hwndDlg,
 
             pGlobalData->NumWavOut = waveOutGetNumDevs();
 
-            SendMessageW(GetDlgItem(hwndDlg, IDC_PLAY_SOUND),
-                         BM_SETIMAGE, (WPARAM)IMAGE_ICON,
-                         (LPARAM)(HANDLE)LoadIconW(hApplet, MAKEINTRESOURCEW(IDI_PLAY_ICON)));
+            pGlobalData->hPlayIcon = LoadImageW(hApplet,
+                                                MAKEINTRESOURCEW(IDI_PLAY_ICON),
+                                                IMAGE_ICON,
+                                                32,
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1234]: hardcoded 32, or `GetSystemMetrics(SM_CXICON)` (and similar for the Y dimension) ?

File: dll/cpl/mmsys/sounds.c [hunk 2]
```diff
@@ -1113,6 +1117,90 @@ InitImageList(UINT StartResource,
...
+ * Returns 0 in case of failure.
+ **/
+static
+ULONG
+GetSoundDuration(LPCWSTR pFileName)
+{
+    DWORD dwSize, dwError;
+    LPWSTR pszCmd;
+    WCHAR Buffer[32];
+    PWCHAR endPtr;
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1138]:
```suggestion
```

File: dll/cpl/mmsys/sounds.c [hunk 3]
```diff
@@ -1113,6 +1117,90 @@ InitImageList(UINT StartResource,
...
+    dwError = mciSendStringW(L"status SoundFile length", Buffer, _countof(Buffer), NULL);
+
+Quit:
+    mciSendStringW(L"close SoundFile", NULL, 0, NULL);
+    HeapFree(GetProcessHeap(), 0, pszCmd);
+
+    if (dwError != 0)
+        return 0;
+
+    return wcstoul(Buffer, &endPtr, 10);
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1167]:
```suggestion
    return wcstoul(Buffer, NULL, 10);
```

File: dll/cpl/mmsys/sounds.c [hunk 4]
```diff
@@ -1140,9 +1228,19 @@ SoundsDlgProc(HWND hwndDlg,
...
+                                                32,
+                                                32,
+                                                LR_DEFAULTCOLOR);
+            pGlobalData->hStopIcon = LoadImageW(hApplet,
+                                                MAKEINTRESOURCEW(IDI_STOP_ICON),
+                                                IMAGE_ICON,
+                                                32,
+                                                32,
+                                                LR_DEFAULTCOLOR);
+            SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON, (LPARAM)pGlobalData->hPlayIcon);
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1243]:
```suggestion
            SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON,
                                (LPARAM)pGlobalData->hPlayIcon);
```

File: dll/cpl/mmsys/sounds.c [hunk 5]
```diff
@@ -1203,6 +1301,13 @@ SoundsDlgProc(HWND hwndDlg,
                 case IDC_PLAY_SOUND:
                 {
                     LRESULT lIndex;
+
+                    if (StopSoundTest(hwndDlg))
+                    {
+                        SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON, (LPARAM)pGlobalData->hPlayIcon);
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1307]:
```suggestion
                        SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON,
                                            (LPARAM)pGlobalData->hPlayIcon);
```

File: dll/cpl/mmsys/sounds.c [hunk 6]
```diff
@@ -1212,7 +1317,10 @@ SoundsDlgProc(HWND hwndDlg,
                     lIndex = ComboBox_GetItemData(GetDlgItem(hwndDlg, IDC_SOUND_LIST), lIndex);
                     if (lIndex != CB_ERR)
                     {
-                        PlaySoundW((PWCHAR)lIndex, NULL, SND_FILENAME);
+                        if (StartSoundTest(hwndDlg, (LPWSTR)lIndex))
+                        {
+                            SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON, (LPARAM)pGlobalData->hStopIcon);
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1322]:
```suggestion
                            SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON,
                                                (LPARAM)pGlobalData->hStopIcon);
```

File: dll/cpl/mmsys/sounds.c [hunk 7]
```diff
@@ -1336,11 +1452,25 @@ SoundsDlgProc(HWND hwndDlg,
                     ApplyChanges(hwndDlg);
                     break;
                 }
+                case PSN_KILLACTIVE:
+                {
+                    if (StopSoundTest(hwndDlg))
+                    {
+                        SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON, (LPARAM)pGlobalData->hPlayIcon);
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1459]:
```suggestion
                        SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON,
                                            (LPARAM)pGlobalData->hPlayIcon);
```

File: dll/cpl/mmsys/sounds.c [hunk 8]
```diff
@@ -1336,11 +1452,25 @@ SoundsDlgProc(HWND hwndDlg,
...
+                    break;
+                }
                 case TVN_SELCHANGED:
                 {
                     LPNMTREEVIEWW nm = (LPNMTREEVIEWW)lParam;
                     LRESULT lCount, lIndex, lResult;
 
+                    if (StopSoundTest(hwndDlg))
+                    {
+                        SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON, (LPARAM)pGlobalData->hPlayIcon);
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1471]:
```suggestion
                        SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON,
                                            (LPARAM)pGlobalData->hPlayIcon);
```

File: dll/cpl/mmsys/sounds.c [hunk 9]
```diff
@@ -1398,8 +1528,18 @@ SoundsDlgProc(HWND hwndDlg,
...
                 }
             }
+            break;
+        }
+        case WM_TIMER:
+        {
+            if (wParam == ID_SOUND_TEST_TIMER)
+            {
+                KillTimer(hwndDlg, ID_SOUND_TEST_TIMER);
+                SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON, (LPARAM)pGlobalData->hPlayIcon);
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1538]:
```suggestion
                SendDlgItemMessageW(hwndDlg, IDC_PLAY_SOUND, BM_SETIMAGE, IMAGE_ICON,
                                    (LPARAM)pGlobalData->hPlayIcon);
```

File: dll/cpl/mmsys/sounds.c [hunk 10]
```diff
@@ -1113,6 +1117,90 @@ InitImageList(UINT StartResource,
...
+    PWCHAR endPtr;
+
+    dwSize = (wcslen(pFileName) + 38 + 1) * sizeof(WCHAR);
+    pszCmd = HeapAlloc(GetProcessHeap(), 0, dwSize);
+    if (!pszCmd)
+        return 0;
+
+    StringCbPrintfW(pszCmd, dwSize, L"open \"%s\" type waveaudio alias SoundFile", pFileName);
+
+    dwError = mciSendStringW(pszCmd, NULL, 0, NULL);
```
Comment (HBelusca) [inline dll/cpl/mmsys/sounds.c:1147]: A question: Did you choose to use those baroque mciSendString commands, in order not to have to open and close an MCI device? (an alternate way would be as we do in mplay32, and use mci_open and mciSendCommand function.)

File: dll/cpl/mmsys/sounds.c (hunk 1 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c:1234]: The ico file has only 32x32 icon, that's why it's hardcoded 32.

File: dll/cpl/mmsys/sounds.c (hunk 2 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c]: Done.

File: dll/cpl/mmsys/sounds.c (hunk 3 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c]: Done.

File: dll/cpl/mmsys/sounds.c (hunk 4 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c]: Done.

File: dll/cpl/mmsys/sounds.c (hunk 5 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c]: Done.

File: dll/cpl/mmsys/sounds.c (hunk 6 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c]: Done.

File: dll/cpl/mmsys/sounds.c (hunk 7 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c]: Done.

File: dll/cpl/mmsys/sounds.c (hunk 8 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c]: Done.

File: dll/cpl/mmsys/sounds.c (hunk 9 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c]: Done.

File: dll/cpl/mmsys/sounds.c (hunk 10 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c:1146]: Why avoid `mciSendString` ? I think it's simpler than `mciSendCommand` in this case.

**Comment by TAN-Gaming** [comment]
Waiting for reviews...

File: dll/cpl/mmsys/sounds.c (hunk 10 above)
Comment (katahiromz) [reply dll/cpl/mmsys/sounds.c:1146]: The strings wastes computing resource. The commands is more computer-friendly.

**Comment by JoachimHenze** [comment]
There is still an unactioned review requests mciSendString->mciSendCommand.
When that will be addressed, I will approve.

File: dll/cpl/mmsys/sounds.c (hunk 10 above)
Comment (TAN-Gaming) [reply dll/cpl/mmsys/sounds.c:1146]: OK.

**Review by katahiromz — APPROVED** [review]

**Review by HBelusca — APPROVED** [review]

**Review by JoachimHenze — APPROVED** [review]
Approved, although I do wonder whether there might be a better solution also, that wouldn't require a timer to restore the icon after the sound finished playing.

**Comment by TAN-Gaming** [comment]
> Approved, although I do wonder whether there might be a better solution also, that wouldn't require a timer to restore the icon after the sound finished playing.

Thanks for the merge. Maybe I'll improve it in the future if I have time.

**Pull request merged**